### ISSUE TITLE
extra verbosity to include request/response bodies

### ISSUE THREAD

**Issue by FND** [cdent/gabbi#69]
Currently it can be somewhat tricky to debug unexpected outcomes, as `verbose: true` only prints headers.

In my case, I wanted to verify that a CSRF token was included in a form submission. The simplest way to check the request body was to start netcat and change my test's URL to http://localhost:9999.

It would be useful if gabbi provided a way to inspect the entire data being sent over the wire.

**Comment by cdent** [comment]
a) At the moment my technique for dealing with this is change the test such that it is a diagnostic, same request info but change status to whatever I'm seeing in real life and check only response_strings with a nonsense value. This will get the error message to show the response body.

b) Obviously this is less than ideal. One potentialy solution would be to change verbose:

* continue to accept `true` to do what it currently does
* also accept other values which signal different levels of verbosity

I'm not sure what would be the best values. Some options for comparison and discussion (no preference on my part should be construed from these representations they are just ideas to move the brain meat):

```yaml
verbose:
    - headers
    - body
```

```yaml
verbose: 2  # or 1, 1 means headers, 2 means headers + body
```

I'm sure there are additional options, such as ways to control if verbose applies to both request and response or only one.

I'd hate to see this get out of hand though. I guess it could also be possible to just turn on full verbose, but the reason I didn't do that is that would get noisy with long bodies (especially binary ones) and I didn't want to get into this situation where only some content-types were visible.

(blah, blah all the usual rules about yeah, good idea, needs to be simple and not noisy and not break the fundamental ambiance)

**Comment by FND** [comment]
I'd also tried (a), but that gets annoying quick.

For the sake of argument, how about this:

```yaml
verbose:
    headers: true
    body: >@/tmp/log
```

(`verbose: true` would continue to mean `verbose.headers: true`, if only for backwards compatibility)

**Comment by cdent** [comment]
Verbosity is already going to STDOUT so (for sake of clarity) why redirect to a file? That is, what need does it satisfy? This may seem obvious to you, but the point is to get that obviousness written down so that the evaluation is as transparent as possible.

Also: that verbose setting is getting pretty verbose itself, which is a huge meh for me because it is adding noise to the test (or the test defaults) which has zero to do with the actual request or the evaluation of the response, and that's what the test is supposed to primarily express. Which is not to say "no way, not gonna happen" but rather: That's the aesthetic guideline you're working against in the effort to come up with a solution and convince of its value.

**Comment by FND** [comment]
It's actually your "that would get noisy with long bodies (especially binary ones)" statement that prompted this idea. Using a file rather than STDOUT would avoid having to identify and extract the body from surrounding output, plus it eliminates the risk ruining your TTY with binary data. It would also allow for easily inspecting multiple bodies in one go by creating separate files. (Obviously any file descriptor should be valid, so you might also use STDERR and redirect that on the shell.)

> that verbose setting is getting pretty verbose itself [...] it is adding noise to the test

Yeah, I didn't want to get into that before, but:

* I'd assumed that `verbose` is almost always used for debugging, i.e. not persisted - thus it would only be temporary and localized noise, which isn't so bad.
* Since the above structures can be annoying to type over and over again during debugging, we'd probably _also_ need a shorthand (with `verbose: 2` being the obvious choice).

Obviously the latter kind of challenges the existence of a verbose structure in the first place. I can't _really_ explain why I like having an explicit version (in particular because there's probably never gonna be a `verbose: 3`, since there's nothing else in HTTP interactions that should be relevant) - except perhaps that it seems easier to learn (first introducing the longhand, then mentioned "oh btw, there's a shorthand too). Of course the shorthand version doesn't allow for custom files either.

**Comment by FND** [comment]
By the way, verbosity should also include the request line (method + URI, including query string and all that).

**Comment by cdent** [comment]
Your syntax seems pretty good but some questions:

If your example above you have:

```yaml
verbose:
    headers: true
    body: >@/tmp/log
```

Is this equivalent to

```yaml
verbose:
    headers: >@/dev/stdout
    body: >@/tmp/log
```

?

Are we imaging that we should allow people to output headers one place and body another? Certainly one can imagine that might be useful, just looking for feedback.

I'm assuming that we would want to open all file for append (so that different tests can append to the same file) and the calling harness/scripts should be responsible for cleaning up log files. Do you concur?

As soon as we start sticking files in like this, it feels inevitable that people will want something like:

```
verbose:
    headers: >@/tmp/$TEST_NAME.log
```

or similar hooey. It's these kinds of slippery slopes that I'm worried about.

Also, you've used `>@`, presumably to be in alignment with the syntax used in `data`, but I'm not sure it is warranted here. `>` can be enough. Or even no symbol at all. `true` and `false` can be special (and in fact are in yaml) and if it isn't one of those it must be a file?

(These sort of mental gyrations are probably why I've avoided this before now)

**Comment by FND** [comment]
> Is [your example] equivalent to [...]

Yeah, pretty much.

> Are we imaging that we should allow people to output headers one place and body another?

Might as well? Seems to a not-uncommon use case with cURL.

> I'm assuming that we would want to open all file for append

Actually - and this is another thing I didn't want to get into before - I was imagining `>>@` was append, `>@` overwrite.

So while, yes, originally that syntax was proposed for consistency (I hadn't even considered your alternative), it also seems more flexible in hindsight. That consistency might also be easier for end users, as targeting files would always work the same way irrespective of context. Your "no symbol at all" approach also makes sense though, so I'm torn.

> the calling harness/scripts should be responsible for cleaning up log files

As far as I'm concerned, gabbi doesn't care whether there's any cleanup at all. Me, I'd probably just write to `/tmp/` and let the OS handle cleanup eventually.

Another thing:

> there's probably never gonna be a `verbose: 3`

Actually, I've since realized that there are things beyond the HTTP wire transmission that might be of interest. In particular, it might be of interest how many times the respective test had to `poll`.

> These sort of mental gyrations are probably why I've avoided this before now

Yeah, it turns out to be far less straightforward than anticipated.

**Comment by cdent** [comment]
On the append issue, my style of use verbose is often to use it for an entire file, so I set it in defaults. In this scenario we wouldn't want overwrite to be too easy to do as that would likely violate the principle of least surprise.

/me continues gyrating

**Comment by jd** [comment]
My point of view, if it can help:

- I don't think numerical value for verbosity level is the way to go in 2015

- Having fine grained on what header and/or body you want from request and/or response with a few options such as `verbose.request.body: true` sounds sane.

- Handling file redirection looks like a rabbit hole you don't want to go into. Let Gabbi have a neat general output that is parseable so anyone can redirect all Gabbi output to a file and then parse and split it as they prefer

- You can try to detect if the body is binary and omit or limit the output

- You can anyway by default limit the ouput to N chars for bodies with an option to control that length, in case someone would need it obviously

**Comment by FND** [comment]
As discussed on IRC, we're considering dropping file redirects, which would allow for a much simpler format:

```yaml
verbose: true
---
verbose: all
---
verbose: headers
---
verbose: body
```

The first two are aliases and would include both headers and body along with any extra info (previously described as `verbose: 3`).

**Comment by cdent** [comment]
I'm doing some local testing with this overly simplistic x.yaml:

```yaml


tests:

    - name: foo1
      url: /

    - name: foo2
      url: /
      verbose: false

    - name: foo3
      url: /
      verbose: true

    - name: foo4
      url: /
      verbose: all

    - name: foo5
      url: /
      verbose: headers

    - name: foo6
      url: /
      verbose: body

    - name: foo7
      method: POST
      url: /
      request_headers:
          content-type: application/json
      data:
          foo: bar
          boo: baz
      verbose: all

    - name: foo8
      method: POST
      url: /
      data:
          foo: bar
          boo: baz
      verbose: body
```

As I said in the commit message this is more to explore the possibilities that to write something in stone.

**Comment by cdent** [comment]
My usual locations for dumping output being dead, here's some sample output here:

```
#### input_foo3 ####
GET /
Host: localhost:8000
> accept-encoding: gzip, deflate
> user-agent: Python-httplib2/0.9.1 (gzip)

< status: 200
< content-location: http://localhost:8000/
< transfer-encoding: chunked
< server: gunicorn/19.3.0
< connection: close
< date: Mon, 10 Aug 2015 18:04:02 GMT
< content-type: text/html; charset=UTF-8

<html>
<head>
    <title>Gabbi Demo Application</title>
</head>
<body>
    <h1>Gabbi Demo Application</h1>

    <p>This is the front page of the Gabbi Demo Application!</p>

    <dl>
        <dt>/</dt>
        <dd>
        <ul>
            <li>GET the front page</li>
            <li>POST to create a container</li>
        </ul>
        </dd>
        <dt>/<em>id</em>
        <dd>
        <ul>
            <li>GET a container</li>
            <li>PUT to update a container</li>
            <li>DELETE a container and its content</li>
        </ul>
        </dd>
        <dt>/<em>id</em>/objects</dt>
        <dd>
        <ul>
            <li>GET a list of objects in a container</li>
            <li>POST to create an object in a container</li>
        </ul>
        </dd>
        <dt>/<em>id</em>/objects/<em>id</em></dt>
        <dd>
        <ul>
            <li>GET an object</li>
            <li>PUT to update an object</li>
            <li>DELETE an object</li>
        </ul>
        </dd>
    </dl>
</body>
</html>


#### input_foo4 ####
GET /
Host: localhost:8000
> accept-encoding: gzip, deflate
> user-agent: Python-httplib2/0.9.1 (gzip)

< status: 200
< content-location: http://localhost:8000/
< transfer-encoding: chunked
< server: gunicorn/19.3.0
< connection: close
< date: Mon, 10 Aug 2015 18:04:02 GMT
< content-type: text/html; charset=UTF-8

<html>
<head>
    <title>Gabbi Demo Application</title>
</head>
<body>
    <h1>Gabbi Demo Application</h1>

    <p>This is the front page of the Gabbi Demo Application!</p>

    <dl>
        <dt>/</dt>
        <dd>
        <ul>
            <li>GET the front page</li>
            <li>POST to create a container</li>
        </ul>
        </dd>
        <dt>/<em>id</em>
        <dd>
        <ul>
            <li>GET a container</li>
            <li>PUT to update a container</li>
            <li>DELETE a container and its content</li>
        </ul>
        </dd>
        <dt>/<em>id</em>/objects</dt>
        <dd>
        <ul>
            <li>GET a list of objects in a container</li>
            <li>POST to create an object in a container</li>
        </ul>
        </dd>
        <dt>/<em>id</em>/objects/<em>id</em></dt>
        <dd>
        <ul>
            <li>GET an object</li>
            <li>PUT to update an object</li>
            <li>DELETE an object</li>
        </ul>
        </dd>
    </dl>
</body>
</html>


#### input_foo5 ####
GET /
Host: localhost:8000
> accept-encoding: gzip, deflate
> user-agent: Python-httplib2/0.9.1 (gzip)

< status: 200
< content-location: http://localhost:8000/
< transfer-encoding: chunked
< server: gunicorn/19.3.0
< connection: close
< date: Mon, 10 Aug 2015 18:04:02 GMT
< content-type: text/html; charset=UTF-8

#### input_foo6 ####
GET /
Host: localhost:8000


<html>
<head>
    <title>Gabbi Demo Application</title>
</head>
<body>
    <h1>Gabbi Demo Application</h1>

    <p>This is the front page of the Gabbi Demo Application!</p>

    <dl>
        <dt>/</dt>
        <dd>
        <ul>
            <li>GET the front page</li>
            <li>POST to create a container</li>
        </ul>
        </dd>
        <dt>/<em>id</em>
        <dd>
        <ul>
            <li>GET a container</li>
            <li>PUT to update a container</li>
            <li>DELETE a container and its content</li>
        </ul>
        </dd>
        <dt>/<em>id</em>/objects</dt>
        <dd>
        <ul>
            <li>GET a list of objects in a container</li>
            <li>POST to create an object in a container</li>
        </ul>
        </dd>
        <dt>/<em>id</em>/objects/<em>id</em></dt>
        <dd>
        <ul>
            <li>GET an object</li>
            <li>PUT to update an object</li>
            <li>DELETE an object</li>
        </ul>
        </dd>
    </dl>
</body>
</html>


#### input_foo7 ####
POST /
Host: localhost:8000
> content-type: application/json
> accept-encoding: gzip, deflate
> user-agent: Python-httplib2/0.9.1 (gzip)

{"foo": "bar", "boo": "baz"}

< status: 500
< connection: close
< content-type: text/html
< content-length: 141

<html>
  <head>
    <title>Internal Server Error</title>
  </head>
  <body>
    <h1><p>Internal Server Error</p></h1>
    
  </body>
</html>


#### input_foo8 ####
POST /
Host: localhost:8000


<html>
  <head>
    <title>Internal Server Error</title>
  </head>
  <body>
    <h1><p>Internal Server Error</p></h1>
    
  </body>
</html>


```

**Comment by cdent** [comment]
Color sample for an example:

![screen shot 2015-08-10 at 20 39 38](https://cloud.githubusercontent.com/assets/34498/9182027/164870c4-3fa0-11e5-9ccf-e9cf41239027.png)

**Comment by FND** [comment]
If you're going with color, might as well highlight header names?

**Comment by cdent** [comment]
At the moment the color handling is per line. It doesn't have to be like that, but the reason it is currently is because there's a separate between message generation and prefix and color handling. If you look at the changes I made [here](https://github.com/cdent/gabbi/commit/9e1172fc83469458da161e19606ece565abe5ef4#diff-28f62ee5a6f319e6efc91dba2322f45bR50) to fix the way you were expressing prefix in your patch, that statement might make slightly more sense.

I agree, however, that the currently color handling is limited. Suggestions welcome.

**Comment by cdent** [comment]
I've gone ahead and merge the verbose-verbose branch and released a 1.5.0 with the increased verbosity control and some color capabilities. It is not perfect, but we can make it so with time and experimentation.
